File: modules/base/test/matlab/AnyTest.m
classdef AnyTest < matlab.unittest.TestCase
    methods (Test)
    
        function defaultConstructor(testCase)
            a = ma.Any;
            testCase.verifyEqual(a.isEmpty(),true);
            testCase.verifyEqual(a.isValid(),false);
            testCase.verifyEqual(a.dimensions(),[]);
            testCase.verifyEqual(a.size(), 0);
            testCase.verifyEqual(a.cast(), 0);
            testCase.verifyEqual(a.cast('int64'), int64(0));
            testCase.verifyEqual(a.cast('char'), '');
            testCase.verifyEqual(a.cast('uint8'), uint8(0));
        end
        
        function otherConstructorSingle(testCase)
            a = ma.Any(1);
            testCase.verifyEqual(a.cast(),1);
            testCase.verifyEqual(a.dimensions(), [1,1]);
            testCase.verifyEqual(a.size(), 1);
            testCase.verifyEqual(a.cast('int64'),int64(1));
            testCase.verifyEqual(a.cast('single'),single(1));
        end
        
        function otherConstructorArray(testCase)
            a = ma.Any([uint64(0) uint64(10) ; uint64(123091278957) uint64(999999)]);
            testCase.verifyEqual(a.cast(),[0, 10 ; 123091278957 999999]);
            testCase.verifyEqual(a.dimensions(), [2,2]);
            testCase.verifyEqual(a.size(), 4);
            testCase.verifyEqual(a.cast('int64'),[int64(0) int64(10) ; int64(123091278957) int64(999999)]);
            testCase.verifyEqual(a.cast('single'),[single(0) single(10) ; single(123091278957) single(999999)]);
            testCase.verifyEqual(a.cast('char'), {'0' '10' ; '123091278957' '999999'});
        end
        
        function otherConstructorLogical(testCase)
            a = ma.Any([true, false, false; false, true, true]);
            testCase.verifyEqual(a.cast(),[1, 0, 0 ; 0, 1, 1]);
            testCase.verifyEqual(a.dimensions(), [2,3]);
            testCase.verifyEqual(a.size(), 6);
            testCase.verifyEqual(a.cast('char'), {'1','0','0' ; '0', '1', '1'});
        end
        
        function otherConstructorCell(testCase)
            a = ma.Any({'foo','bar'});
            testCase.verifyEqual(a.cast(),[0, 0]);
            testCase.verifyEqual(a.dimensions(), [1,2]);
            testCase.verifyEqual(a.size(), 2);
            testCase.verifyEqual(a.cast('char'), {'foo','bar'});
        end
        
    end
end

File: modules/base/test/python/anyTest.py
import ma
import unittest

class AnyTest(unittest.TestCase):
    def test_default_ctor(self):
        a = ma.Any()
        self.assertEqual(a.isEmpty(), True)
        self.assertEqual(a.isValid(), False)
        self.assertEqual(a.dimensions(),[])
        self.assertEqual(a.size(), 0)
        self.assertEqual(a.cast(), 0.)
        self.assertEqual(a.cast('bool'), False)
        self.assertEqual(a.cast('int'), 0)
        self.assertEqual(a.cast('long'), 0)
        self.assertEqual(a.cast('string'), '')
        
    def test_other_constructor_single(self):
        a = ma.Any(1)
        self.assertEqual(a.cast(),1.)
        self.assertEqual(a.dimensions(), [])
        self.assertEqual(a.size(), 1)
        self.assertEqual(a.cast('int'),1)
        self.assertEqual(a.cast('float'),1.)

    def test_other_constructor_array(self):
        a = ma.Any([[0, 10], [12309127, 999999]])
        self.assertEqual(a.dimensions(), [2,2])
        self.assertEqual(a.size(), 4)
        self.assertEqual(a.cast(),[[0., 10.], [12309127., 999999.]])
        self.assertEqual(a.cast('long'), [[0, 10], [12309127, 999999]])
        self.assertEqual(a.cast('string'), [['0', '10'], ['12309127', '999999']])

    def test_other_constructor_logical(self):
        a = ma.Any([[True, False, False], [False, True, True]]);
        self.assertEqual(a.dimensions(), [2,3]);
        self.assertEqual(a.size(), 6);
        self.assertEqual(a.cast(),[[1., 0., 0.], [0, 1, 1]]);
        self.assertEqual(a.cast('bool'),[[True, False, False], [False, True, True]]);
        self.assertEqual(a.cast('string'), [['1','0','0'], ['0', '1', '1']]);

    def test_other_constructor_List(self):
        a = ma.Any(['foo','bar']);
        self.assertEqual(a.dimensions(), [2]);
        self.assertEqual(a.size(), 2);
        self.assertEqual(a.cast(),[0, 0]);
        self.assertEqual(a.cast('string'), ['foo','bar']);
    
    def test_unsupported_type(self):
        self.assertRaises(TypeError, ma.Any, 1j)
        a = ma.Any()
        self.assertRaises(TypeError, a.assign, 1j)
        
    def test_mix_type(self):
        self.assertRaises(TypeError, ma.Any, [])
        self.assertRaises(TypeError, ma.Any, [[],[]])
        self.assertRaises(TypeError, ma.Any, [[],''])
        self.assertRaises(TypeError, ma.Any, ['',[]])
        self.assertRaises(TypeError, ma.Any, [1.,1])
        self.assertRaises(TypeError, ma.Any, [1, 1j])
        self.assertRaises(TypeError, ma.Any, [1, 1, 1j])
        self.assertRaises(TypeError, ma.Any, [1,'1'])
        self.assertRaises(TypeError, ma.Any, [[1],['1']])
        self.assertRaises(TypeError, ma.Any, [[1, 1.],['1','1']])
        self.assertRaises(TypeError, ma.Any, [1,[1,1]])
        self.assertRaises(TypeError, ma.Any, [[1, 1],'1'])
        self.assertRaises(TypeError, ma.Any, [[1, 1],[1.,'1']])
        self.assertRaises(TypeError, ma.Any, [[[1],[1]],[[1.],['1']]])
        
    def test_mix_dims(self):
        self.assertRaises(ValueError, ma.Any, [[1,1],[1,1,1]])